components/renku_data_services/session/db.py

```diff
@@ -89,6 +89,37 @@ def __insert_environment(
         session.add(environment)
         return environment
 
+    def __copy_environment(
+        self,
+        user: base_models.APIUser,
+        session: AsyncSession,
+        environment: models.Environment,
+    ) -> schemas.EnvironmentORM:
+        if user.id is None:
+            raise errors.UnauthorizedError(
+                message="You have to be authenticated to insert an environment in the DB.", quiet=True
+            )
+        new_environment = schemas.EnvironmentORM(
+            name=environment.name,
+            created_by_id=user.id,
+            description=environment.description,
+            container_image=environment.container_image,
+            default_url=environment.default_url,
+            port=environment.port,
+            working_directory=environment.working_directory,
+            mount_directory=environment.mount_directory,
+            uid=environment.uid,
+            gid=environment.gid,
+            environment_kind=environment.environment_kind,
+            command=environment.command,
+            args=environment.args,
+            creation_date=datetime.now(UTC).replace(microsecond=0),
```

Comment:
Non-blocking note: why do we even do this `.replace(microsecond=0)`? This is not useful and the ULID will already contain millisecond-precision information.

Reply:
I have no clue @leafty I have seen it in a few places in the code and wondered myself.

Reply:
https://github.com/SwissDataScienceCenter/renku-data-services/pull/175/commits/fbeae73ba0832bf7b4f53dbd0e8fbffe74914603#diff-e6ad745574f64b4241525aaa0d39f7736545db13f97f9092f00b723257970b9cR228 is the first occurrence of it that I could find but it doesn't offer any clue on why it was added.